src/core/utilities/sorting.hpp

```diff
@@ -0,0 +1,78 @@
+#ifndef PHARE_SORTING_HPP
+#define PHARE_SORTING_HPP
+
+#include "core/utilities/box/box.hpp"
+
+#include <iostream>
+#include <cstddef>
+#include <vector>
+
+namespace PHARE::core
+{
+
+template<typename Array, std::size_t dim>
+class CountingSort
+{
+public:
+    CountingSort(std::size_t size = 0)
+        : tmp_(size)
+    {
+    }
+
+    using value_type = typename Array::value_type;
+    using SortBox    = core::Box<int, dim>;
+
+    void setup(std::size_t nbr_elements, SortBox const& box)
```

Comment:
I've put the allocations in a separate routine so it can be done elsewhere and not occur when sorting.
It's not great yet since the box given here to allocate the histogram must be the same as the one used by the cellGetter passed to sort() otherwise it's shit.

particle array given to sort() must also be consistent with what's been given as `nbr_elements` otherwise nonsense... 

TBC